src/charm.py

```diff
@@ -1,104 +1,200 @@
-            self._stored.things.append(current)
-
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
-
-        Learn more about actions at https://juju.is/docs/sdk/actions
-        """
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
-        else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
-
-
-if __name__ == "__main__":
-    main(OperatorTemplateCharm)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on.leader_settings_changed, self._on_leader_settings_changed)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        self.framework.observe(self.on.secret_changed, self._on_secret_changed)
+
+        # resource patching
+        self.framework.observe(
+            self.resources_patch.on.patch_failed, self._on_resource_patch_failed
+        )
+
+        # ingress
+        self.framework.observe(
+            self.ingress.on.ready,
+            self._on_ingress_changed,
+        )
+
+    @property
+    def _pebble_layer(self) -> ops.pebble.Layer:
+        return self._pebble_service.render_pebble_layer(
+            TracingData.load(self.tracing_requirer),
+            self._secrets,
+            self._config,
+        )
+
+    @leader_unit
+    def _prepare_secrets(self) -> None:
+        self._secrets[API_TOKEN_SECRET_LABEL] = {API_TOKEN_SECRET_KEY: token_hex(16)}
+
+    @leader_unit
+    def _on_ingress_changed(self, event: ops.RelationEvent) -> None:
+        if self.ingress.is_ready():
+            ingress_config = IngressData.load(self.ingress).config
+            self.ingress.submit_to_traefik(ingress_config)
+        self._holistic_handler(event)
+
+    def _on_leader_elected(self, event: ops.LeaderElectedEvent) -> None:
+        self._holistic_handler(event)
+
+    def _on_leader_settings_changed(self, event: ops.LeaderSettingsChangedEvent) -> None:
+        self._holistic_handler(event)
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        self._holistic_handler(event)
+
+    def _on_secret_changed(self, event: ops.SecretChangedEvent) -> None:
+        self._holistic_handler(event)
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent):
+        self._workload_service.open_port()
+        self._holistic_handler(event)
+
+        self._workload_service.set_version()
+
+    def _on_resource_patch_failed(self, event: K8sResourcePatchFailedEvent) -> None:
+        logger.error(f"Failed to patch resource constraints: {event.message}")
+        self.unit.status = ops.BlockedStatus(event.message)
```

Comment:
q: If I understand correctly, this blocked status can be overwritten by the holistic handler if some other event is fired?

Reply:
that's fine, let's keep it as is